includes/class-admin-plugins-screen.php

```diff
@@ -0,0 +1,234 @@
+			}
+		}
+
+		return $plugins;
+	}
+
+	/**
+	 * Remove 'Activate' and 'Delete' links for uninstalled plugins, and add an 'Install' link.
+	 *
+	 * @param  array  $actions Array of plugin action links.
+	 * @param  string $plugin_file The plugin file.
+	 * @param  array  $plugin_data Information about the plugin.
+	 * @return array  Modified $actions.
+	 */
+	public function modify_action_links( $actions, $plugin_file, $plugin_data ) {
+		$plugin_slug       = isset( $plugin_data['slug'] ) ? $plugin_data['slug'] : $plugin_file;
+		$installed_plugins = Plugin_Manager::get_installed_plugins();
+		$managed_plugins   = array_keys( Plugin_Manager::get_managed_plugins() );
+
+		if ( in_array( $plugin_slug, $managed_plugins ) && ! isset( $installed_plugins[ $plugin_slug ] ) ) {
+			unset( $actions['activate'] );
+			unset( $actions['delete'] );
+			if ( current_user_can( 'install_plugins' ) ) {
+				$actions['install'] = '<a href="' .
+					wp_nonce_url( 'plugins.php?action=newspack_install_plugin&plugin=' . urlencode( $plugin_slug ), 'newspack-install-plugin_' . $plugin_slug, 'install_nonce' ) .
+					'" class="edit" aria-label="' .
+					/* translators: %s - plugin name */
+					esc_attr( sprintf( __( 'Install %s', 'newspack' ), $plugin_data['Name'] ) ) .
+					'">' .
+					__( 'Install', 'newspack' ) .
+					'</a>';
+			}
+		}
+
+		return $actions;
+	}
+
+	/**
+	 * Install a plugin when the installation link is clicked.
+	 */
+	public function handle_plugin_install() {
+		if ( ! current_user_can( 'install_plugins' ) ) {
+			wp_die( esc_html__( 'Sorry, you are not allowed to install plugins.', 'newspack' ) );
+		}
+
+		$plugin = isset( $_REQUEST['plugin'] ) ? sanitize_title( $_REQUEST['plugin'] ) : ''; // phpcs:ignore WordPress.VIP.SuperGlobalInputUsage
+		if ( ! $plugin ) {
+			wp_die( esc_html__( 'Invalid plugin.', 'newspack' ) );
+		}
+		check_admin_referer( 'newspack-install-plugin_' . $plugin, 'install_nonce' );
+
+		$managed_plugins = Plugin_Manager::get_managed_plugins();
+		if ( ! isset( $managed_plugins[ $plugin ] ) ) {
+			wp_die( esc_html__( 'Invalid plugin.', 'newspack' ) );
+		}
+
+		$plugin_info = $managed_plugins[ $plugin ];
+		if ( 'wporg' === $plugin_info['Download'] ) {
+			$result = Plugin_Manager::install( $plugin );
+		} else {
+			$result = Plugin_Manager::install( $plugin_info['Download'] );
+		}
+
+		if ( is_wp_error( $result ) ) {
+			$redirect = self_admin_url( 'plugins.php?newspack_install_error=true&message=' . urlencode( $result->get_error_message() ) . '&plugin=' . urlencode( $plugin ) );
+			wp_safe_redirect( add_query_arg( '_error_nonce', wp_create_nonce( 'newspack-install-error_' . $plugin ), $redirect ) );
+			exit;
+		} else {
+			wp_safe_redirect( self_admin_url( 'plugins.php?newspack_install_success=true' ) );
+			exit;
+		}
+	}
+
+	/**
+	 * Output success/failure notices on plugin installation.
+	 */
+	public function plugin_install_notices() {
```

Comment:
Given all the PHPCS ignore rules I was looking this over trying to see if there was any other way to go. First question: why do we need this hook at all? I'm unclear on the reason for having custom notices? Second, I was wondering if there's a way to use `[filter_input](https://www.php.net/manual/en/function.filter-input.php)` to get `newspack_install_success` from the `$_REQUEST` superglobal without needing the ignores?

Reply:
1. The notices are displayed when a plugin is installed (either success/failure notice). We need to use custom ones since installation of plugins through the plugin management screen isn't a feature in regular WP, so there isn't a non-custom one available. [These notices are implemented in the standard way.](https://premium.wpmudev.org/blog/adding-admin-notices/)

2. `filter_input` prevented the need for the phpcs ignores. Good tip!